src/core/transaction.c

```diff
@@ -902,6 +902,7 @@ int transaction_add_job_and_dependencies(
                 bool ignore_order,
                 sd_bus_error *e) {
 
+        _cleanup_strv_free_ char **propagated_restart = NULL;
```

Comment:
string arrays suck for this. Please use a `Set` object, and just insert the `Unit` object pointer into it.

This will bring down rutime behaviour for checking if we already marked something for restarts from O(n) to effectively O(1)

Reply:
also, a global field for this sounds like a strange option to me. this should be moved into the loop immediately before the two dep loops.

Reply:
> also, a global field for this sounds like a strange option to me. this should be moved into the loop immediately before the two dep loops.

Hmm, I didn't seem to find such a loop 🤔

Reply:
Does declaring variables in `if` blocks meet the latest code style? It seems that this piece of code hasn't been touched for ages.